core/iwasm/compilation/aot_emit_memory.c

```diff
@@ -286,10 +318,25 @@ aot_check_memory_overflow(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
                        MEMORY64_COND_VALUE(I64_ZERO, I32_ZERO), cmp, "is_zero");
             ADD_BASIC_BLOCK(check_succ, "check_mem_size_succ");
             LLVMMoveBasicBlockAfter(check_succ, block_curr);
-            if (!aot_emit_exception(comp_ctx, func_ctx,
-                                    EXCE_OUT_OF_BOUNDS_MEMORY_ACCESS, true, cmp,
-                                    check_succ)) {
-                goto fail;
+
+            if (!comp_ctx->enable_runtime_bound_check) {
+                if (!aot_emit_exception(comp_ctx, func_ctx,
+                                        EXCE_OUT_OF_BOUNDS_MEMORY_ACCESS, true,
+                                        cmp, check_succ)) {
+                    goto fail;
+                }
+            }
+            else {
+                ADD_BASIC_BLOCK(bound_check_with_api, "bound_check_with_api");
+                ADD_BASIC_BLOCK(check_end, "check_end");
+                LLVMMoveBasicBlockAfter(bound_check_with_api, check_succ);
+                LLVMMoveBasicBlockAfter(check_end, bound_check_with_api);
+
+                if (!LLVMBuildCondBr(comp_ctx->builder, cmp,
```

Comment:
The main issue now is that the additional conditional jumps themselves cause performance problems, whether using runtime APIs or directly performing boundary checks with LLVM IR.